### PR TITLE
Address all remaining clang-tidy errors

### PR DESCRIPTION
**Pull request by vyasr**
## Description

With this set of changes I get a clean run of clang-tidy (with one caveat that I'll explain in the follow-up PR to add clang-tidy to pre-commit/CI).

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/tests/transform/bools_to_mask_test.cpp
Comment (davidwendt) [inline cpp/tests/transform/bools_to_mask_test.cpp:35]: What? Um wait, what?

Comment (vyasr) [reply cpp/tests/transform/bools_to_mask_test.cpp:35]: This is https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-transparent-functors.html, which is because in C++14 the template arguments for so-called "transparent" functors can be deduced in the right contexts. In this case there isn't really any benefit though and we could go ahead and revert. I believe the primary reason that this functionality was added was to supported heterogeneous lookup when the functors needed to operate on different types. Up to you.

File: cpp/tests/structs/structs_column_tests.cpp
Comment (vyasr) [reply cpp/tests/structs/structs_column_tests.cpp:640]: I had to capture the returned value because `view` is marked as nodiscard. Presumably the test is just to make sure that view construction succeeds, is that what you're asking? We could also tell the linter to ignore this one.

File: cpp/tests/transform/bools_to_mask_test.cpp
Comment (vyasr) [reply cpp/tests/transform/bools_to_mask_test.cpp:35]: [This thread](https://stackoverflow.com/questions/20317413/what-are-transparent-comparators) has some good info.

Comment (davidwendt) [reply cpp/tests/transform/bools_to_mask_test.cpp:35]: I learned about transparent functors today. I'm good with this change. Thanks.

File: cpp/tests/structs/structs_column_tests.cpp
Comment (davidwendt) [reply cpp/tests/structs/structs_column_tests.cpp:640]: So this is just testing the `view()` member of `cudf::column` does not throw an exception which likely will not every happen: https://github.com/rapidsai/cudf/blob/9b2f892c5ec59605bfdc3a2abe4885176950589a/cpp/src/column/column.cu#L89
I'm guessing this was meant to check `cudf::make_structs_column` does not throw an exception which mostly definitely can throw:
https://github.com/rapidsai/cudf/blob/9b2f892c5ec59605bfdc3a2abe4885176950589a/cpp/src/structs/structs_column_factories.cu#L31
So I think 
```
  EXPECT_NO_THROW(auto structs_col =
    cudf::make_structs_column(num_rows, std::move(cols), null_count, std::move(null_mask)));
```
would be more appropriate. And remove the check for the `view()` call.

Comment (vyasr) [reply cpp/tests/structs/structs_column_tests.cpp:640]: Makes sense, I'll do that.

**Review by ttnghia — APPROVED** [review]
Personally I don't like the idea of removing the name class when constructing an object. Having the name class makes thing looks more obvious and less error prone. In particular, instead of having this:
```
class_name func() {
  // ... very long code
  return {a, b, c,...};
}
```
I prefer to have this:
```
class_name func() {
  // ... very long code
  return class_name{a, b, c,...};
}

```

**Comment by davidwendt** [comment]
> Personally I don't like the idea of removing the name class when constructing an object. Having the name class makes thing looks more obvious and less error prone. In particular, instead of having this:
> 
> ```
> class_name func() {
>   // ... very long code
>   return {a, b, c,...};
> }
> ```
> 
> I prefer to have this:
> 
> ```
> class_name func() {
>   // ... very long code
>   return class_name{a, b, c,...};
> }
> ```

I agree with this as well but will abide by a majority vote.

File: cpp/src/io/parquet/compact_protocol_reader.cpp
Comment (mythrocks) [inline cpp/src/io/parquet/compact_protocol_reader.cpp:231]: ELI5:  Why was this change required?  Isn't the old version a valid equivalent?

Is it because of (the equivalent of) the [Google Style Guide](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) stating that `structs` need to be POD/passive objects?

File: cpp/src/utilities/host_memory.cpp
Comment (mythrocks) [inline cpp/src/utilities/host_memory.cpp:122]: > If cudf were _ever to try_ to support clang...

You didn't split that infinitive. 🎩 Good show, there. :] (I know, it shouldn't matter.)

File: cpp/tests/binaryop/binop-compiled-test.cpp
Comment (mythrocks) [inline cpp/tests/binaryop/binop-compiled-test.cpp:564]: I'm not sure I've understood the need for this either.  Why is repeating the variable name here preferred?

File: cpp/tests/copying/slice_tests.cuh
Comment (mythrocks) [inline cpp/tests/copying/slice_tests.cuh:151]: 👍

File: cpp/tests/binaryop/binop-compiled-test.cpp
Comment (ttnghia) [reply cpp/tests/binaryop/binop-compiled-test.cpp:564]: Structure binding cannot be captured automatically so we have to use this trick.

File: cpp/tests/io/comp/decomp_test.cpp
Comment (mythrocks) [inline cpp/tests/io/comp/decomp_test.cpp:42]: Nit:  Why not `std::string const& str`?  It's not like `str` is being `std::move()`-ed.

File: cpp/tests/io/csv_test.cpp
Comment (mythrocks) [inline cpp/tests/io/csv_test.cpp:957]: 👍

**Review by mythrocks — APPROVED** [review]
A couple of nits and questions.  But this LGTM.

File: cpp/tests/binaryop/binop-compiled-test.cpp
Comment (mythrocks) [reply cpp/tests/binaryop/binop-compiled-test.cpp:564]: > Structure binding cannot be captured automatically...

In which case then, how was it working before?  The definitions of `lhs_data` etc. haven't changed, right?

**Comment by mythrocks** [comment]
>> Personally I don't like the idea of removing the name class when constructing an object. 

> I agree with this as well.

To me it's a little bit like using `auto` to define variables for objects returned from a function, whose type isn't immediately evident at the call-site.  E.g. `std::size_t` vs `cudf::size_type`.  

I think I'd prefer the explicit type for clarity, but I'll follow what the team decides.

File: cpp/tests/binaryop/binop-compiled-test.cpp
Comment (vyasr) [reply cpp/tests/binaryop/binop-compiled-test.cpp:564]: We were getting lucky because gcc implements this as an extension even when only compiling under C++17, but other compilers will not be so friendly: https://godbolt.org/z/9fnq6Ex84.

File: cpp/src/io/parquet/compact_protocol_reader.cpp
Comment (vyasr) [reply cpp/src/io/parquet/compact_protocol_reader.cpp:231]: No, this was necessary because the header file declares this object as a class instead of a struct. We could equivalently have changed the header to use struct instead, but the two should be aligned because under certain compilers (Windows) the results could be ABI-incompatible. See https://clang.llvm.org/docs/DiagnosticsReference.html#wmismatched-tags

File: cpp/tests/io/comp/decomp_test.cpp
Comment (vyasr) [reply cpp/tests/io/comp/decomp_test.cpp:42]: Good question. I can make this change.

**Comment by vyasr** [comment]
> > > Personally I don't like the idea of removing the name class when constructing an object.
> 
> > I agree with this as well.
> 
> To me it's a little bit like using `auto` to define variables for objects returned from a function, whose type isn't immediately evident at the call-site. E.g. `std::size_t` vs `cudf::size_type`.
> 
> I think I'd prefer the explicit type for clarity, but I'll follow what the team decides.

I agree with the auto analogy. Even more than Mithun's example, this is the exact mirror of using an `auto` return type from a function; in that case the signature doesn't have the type because it is inferred from the code, whereas in the case we're discussing here the type is only in the signature and the code grabs it from there. I suspect this preference will also have a lot to do with what other languages a given developer is familiar with.

Personally I like the change, but I don't feel strongly enough to sway the team if other generally prefer the status quo. It seems like the vote is currently 3-1 so I'll go ahead and revert. I'll need to go through and manually revert the changes after removing the rule since these changes are not autofixes, so I'll come back to this tomorrow.

**Comment by davidwendt** [comment]
> > > > Personally I don't like the idea of removing the name class when constructing an object.
> > 
> > 
> > > I agree with this as well.
> > 
> > 
> > To me it's a little bit like using `auto` to define variables for objects returned from a function, whose type isn't immediately evident at the call-site. E.g. `std::size_t` vs `cudf::size_type`.
> > I think I'd prefer the explicit type for clarity, but I'll follow what the team decides.
> 
> I agree with the auto analogy. Even more than Mithun's example, this is the exact mirror of using an `auto` return type from a function; in that case the signature doesn't have the type because it is inferred from the code, whereas in the case we're discussing here the type is only in the signature and the code grabs it from there. I suspect this preference will also have a lot to do with what other languages a given developer is familiar with.
> 
> Personally I like the change, but I don't feel strongly enough to sway the team if other generally prefer the status quo. It seems like the vote is currently 3-1 so I'll go ahead and revert. I'll need to go through and manually revert the changes after removing the rule since these changes are not autofixes, so I'll come back to this tomorrow.

@bdice @shrshi @vuule @PointKernel @karthikeyann @mhaseeb123 @kingcrimsontianyu @lamarr

**Comment by vuule** [comment]
I like using `auto` in almost every context. It prevents accidental conversions and my IDE can tell me what type is actually used. I've had to deal with multiple bugs caused by conversions on return, but I don't think I've seen any bugs caused by `auto` use.

I do understand the annoyance with long functions that return "auto", so I'm not really pushing for such `auto` use. Fine either way, just wanted to highlight what I see as a benefit.

**Comment by PointKernel** [comment]
I prefer the explicit return type for better clarity and readability as well.

Also, given that many of our unary constructors are not marked as explicit but should be, using explicit return types is required in such cases.

**Comment by bdice** [comment]
I prefer `return class_name{a, b, c,...};` over `return {a, b, c, …};`.

I am fine with using `auto` in a function’s declared return type like `auto function_name(){}`, especially when paired with the explicit class name at the return statement having some friendly CTAD simplification.

**Comment by vyasr** [comment]
I think that is enough votes that I will go ahead and revert this change and remove the relevant clang-tidy rule.

**Comment by vyasr** [comment]
I have reverted the change. @davidwendt could you take another look and see if there's anything else that stands out? This PR has two approvals but given the discussion and subsequent changes I'd like another pair of eyes on it.

**Comment by ttnghia** [comment]
With the config removed, can we use braces instead of parenthesis in the constructor call? I.e., instead of `class_name(...)` we use `class_name{...}`?

**Comment by vyasr** [comment]
We can, but we have to do it manually. You can't have clang-tidy do that automatically for you. As such, I would prefer not to make such changes in this PR because they're a stylistic improvement that's out of scope of what clang-tidy can do.

**Review by davidwendt — APPROVED** [review]

**Review by KyleFromNVIDIA — APPROVED** [review]
Approved CMake changes

**Pull request merged**
